contrib/hashicorp/consul/consul.go

```diff
@@ -0,0 +1,162 @@
+package consul
+
+import (
+	"context"
+	"math"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+
+	consul "github.com/hashicorp/consul/api"
+)
+
+// Client wraps the regular *consul.Client and augments it with tracing. Use NewClient to initialize it.
+type Client struct {
+	*consul.Client
+
+	config *clientConfig
+	ctx    context.Context
+}
+
+// NewClient returns a traced Consul client.
+func NewClient(config *consul.Config, opts ...ClientOption) (*Client, error) {
+	c, err := consul.NewClient(config)
+	if err != nil {
+		return nil, err
+	}
+	return WrapClient(c, opts...), nil
+}
+
+// WrapClient wraps a given consul.Client with a tracer under the given service name.
+func WrapClient(c *consul.Client, opts ...ClientOption) *Client {
+	cfg := new(clientConfig)
+	defaults(cfg)
+	for _, fn := range opts {
+		fn(cfg)
+	}
+	return &Client{c, cfg, context.Background()}
+}
+
+// WithContext sets a context on a Client. Use it to ensure that emitted spans have the correct parent.
+func (c *Client) WithContext(ctx context.Context) *Client {
+	c.ctx = ctx
+	return c
+}
+
+// A KV is used to trace requests to Consul's KV.
+type KV struct {
+	*consul.KV
+
+	config *clientConfig
+	ctx    context.Context
+}
+
+// KV returns the KV for the Client.
+func (c *Client) KV() *KV {
+	return &KV{c.Client.KV(), c.config, c.ctx}
+}
+
+func (k *KV) startSpan(resourceName string, key string) ddtrace.Span {
+	opts := []ddtrace.StartSpanOption{
+		tracer.ResourceName(resourceName),
+		tracer.ServiceName(k.config.serviceName),
+		tracer.SpanType(ext.SpanTypeConsul),
+		tracer.Tag("consul.key", key),
+	}
+	if !math.IsNaN(k.config.analyticsRate) {
+		opts = append(opts, tracer.Tag(ext.EventSampleRate, k.config.analyticsRate))
+	}
+	span, _ := tracer.StartSpanFromContext(k.ctx, k.config.spanName, opts...)
+	return span
+}
+
+// Put is used to write a new value. Only the
+// Key, Flags and Value is respected.
+func (k *KV) Put(p *consul.KVPair, q *consul.WriteOptions) (*consul.WriteMeta, error) {
+	span := k.startSpan("PUT", p.Key)
+	meta, err := k.KV.Put(p, q)
+	defer span.Finish(tracer.WithError(err))
+	return meta, err
+}
+
+// Get is used to lookup a single key. The returned pointer
+// to the KVPair will be nil if the key does not exist.
+func (k *KV) Get(key string, q *consul.QueryOptions) (*consul.KVPair, *consul.QueryMeta, error) {
+	span := k.startSpan("GET", key)
```

Comment:
I don't know if this is better or not, but should we have higher cardinality here? For example, in our HTTP integrations, we don't use only `GET` as the resource name, but also the path pattern, e.g. `GET /user/{id}`. Should we do the same here? I think it would be nicer to do:
```go
resource := "GET " + key
span := k.startSpan("GET " + key, key)
```
And the same for the others (`LIST`, `KEYS`, `CAS`, `ACQUIRE`, ...). In which case, maybe we can also remove the `consul.key` tag.

What are your thoughts on this?

Reply:
After discussion, I've decided to keep just the operation name as the span name, and keep the path in a tag. This is the safest option in terms of keeping cardinality low while keeping all information accessible.